src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventTypesAnalyzer.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage.Events {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class EventTypesAnalyzer : DiagnosticAnalyzer {
+
+		private const string EventAttributeFullName = "D2L.LP.Distributed.Events.Domain.EventAttribute";
+		private const string ImmutableAttributeFullName = "D2L.CodeStyle.Annotations.Objects+ImmutableAttribute";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.EventTypeMissingImmutableAttribute
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			if( !compilation.TryGetTypeByMetadataName( EventAttributeFullName, out INamedTypeSymbol eventAttributeType ) ) {
+				return;
+			}
+
+			INamedTypeSymbol immutableAttributeType = compilation.GetTypeByMetadataName( ImmutableAttributeFullName );
+
+			context.RegisterSyntaxNodeAction(
+					ctxt => AnalyzeMethodInvocation(
+						ctxt,
+						(ClassDeclarationSyntax)ctxt.Node,
+						eventAttributeType,
+						immutableAttributeType
+					),
+					SyntaxKind.ClassDeclaration
+				);
+		}
+
+		private void AnalyzeMethodInvocation(
+				SyntaxNodeAnalysisContext context,
+				ClassDeclarationSyntax declaration,
+				INamedTypeSymbol eventAttributeType,
+				INamedTypeSymbol immutableAttributeType
+			) {
+
+			bool hasEventAttribute = HasAttribute( context, declaration, eventAttributeType );
+			if( !hasEventAttribute ) {
+				return;
+			}
+
+			bool hasImmutableAttirbute = HasAttribute( context, declaration, immutableAttributeType );
+			if( hasImmutableAttirbute ) {
+				return;
+			}
+
+			INamedTypeSymbol classSymbol = context.SemanticModel.GetDeclaredSymbol( declaration );
+
+			Diagnostic diagnostic = Diagnostic.Create(
+					Diagnostics.EventTypeMissingImmutableAttribute,
+					declaration.Identifier.GetLocation(),
+					classSymbol.ToDisplayString()
+				);
+
+			context.ReportDiagnostic( diagnostic );
+		}
+
+		private static bool HasAttribute(
+				SyntaxNodeAnalysisContext context,
+				ClassDeclarationSyntax declaration,
+				INamedTypeSymbol attributeType
+			) {
+
+			if( attributeType == null ) {
+				return false;
+			}
+
+			foreach( AttributeListSyntax attrList in declaration.AttributeLists ) {

```

Comment:
@j3parker ,  Do you know of a better way to get all attributes for given ITypeSymbol?

Reply:
Yep, going through semantic objects (vs. syntactic) like `ITypeSymbol` will work better. Here's an example:

https://github.com/Brightspace/D2L.CodeStyle/blob/master/src/D2L.CodeStyle.Analyzers/Immutability/ExplicitImmutableAttributeTransitivityAnalyzer.cs#L47-L49

Reply:
Sorry, linked to the wrong lines. Fixed.

Reply:
Thanks